k4MarlinWrapper/src/components/EDM4hep2Lcio.cpp

```diff
@@ -302,14 +310,59 @@ podio::CollectionBase* EDM4hep2LcioTool::getEDM4hepCollection(const std::string&
   throw GaudiException("Collection could not be casted to the expected type", name(), StatusCode::FAILURE);
 }
 
+StatusCode EDM4hep2LcioTool::getAvailableCollectionsFromStore() {
+  SmartIF<IDataManagerSvc> mgr;
+  mgr = evtSvc();
+
+  SmartDataPtr<DataObject> root(evtSvc(), "/Event");
+  if (!root) {
+    error() << "Failed to retrieve root object /Event" << endmsg;
+    return StatusCode::FAILURE;
+  }
+
+  auto pObj = root->registry();
+  if (!pObj) {
+    error() << "Failed to retrieve the root registry object" << endmsg;
+    return StatusCode::FAILURE;
+  }
+  std::vector<IRegistry*> leaves;
+  StatusCode              sc = mgr->objectLeaves(pObj, leaves);
+  if (!sc.isSuccess()) {
+    error() << "Failed to retrieve object leaves" << endmsg;
+    return StatusCode::FAILURE;
+  }
+  for (const auto& pReg : leaves) {
+    if (pReg->name() == k4FWCore::frameLocation) {
+      continue;
+    }
+    DataObject* p;
+    sc = m_eventDataSvc->retrieveObject("/Event" + pReg->name(), p);
+    if (sc.isFailure()) {
+      error() << "Could not retrieve object " << pReg->name() << " from the EventStore" << endmsg;
+      return sc;
+    }
+    auto wrapper = dynamic_cast<AnyDataWrapper<std::unique_ptr<podio::CollectionBase>>*>(p);
+    if (!wrapper) {
+      continue;
+    }
+    // Remove the leading /
+    m_collectionNames.push_back(pReg->name().substr(1, pReg->name().size() - 1));
+    m_idToName.emplace(wrapper->getData()->getID(), pReg->name());
+  }
+  return StatusCode::SUCCESS;
+}
+
 // Select the appropriate method to convert a collection given its type
 void EDM4hep2LcioTool::convertAdd(const std::string& e4h_coll_name, const std::string& lcio_coll_name,
                                   lcio::LCEventImpl* lcio_event, CollectionPairMappings& collection_pairs,
                                   std::vector<EDM4hep2LCIOConv::ParticleIDConvData>& pidCollections,
                                   std::vector<EDM4hep2LCIOConv::TrackDqdxConvData>&  dQdxCollections) {
-  const auto& metadata = m_podioDataSvc->getMetaDataFrame();
-  const auto  collPtr  = getEDM4hepCollection(e4h_coll_name);
-  const auto  fulltype = collPtr->getValueTypeName();
+  std::optional<std::reference_wrapper<podio::Frame>> metadata;
+  if (m_podioDataSvc) {
+    metadata = m_podioDataSvc->getMetaDataFrame();
+  }
```

Comment:
```suggestion
```
I think this is unnecessary because `metadata` is only used behind the exact same check further below. Hence, we can simply get the Frame there without having to park it into an optional here.

Reply:
Yes, this is not needed